processor/groupbytraceprocessor/storage_memory.go

```diff
@@ -0,0 +1,101 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package groupbytraceprocessor
+
+import (
+	"errors"
+	"sync"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+var errNilResourceSpans = errors.New("the provided trace is invalid (nil)")
+
+type memoryStorage struct {
+	sync.RWMutex
+	content map[string]pdata.ResourceSpans
+}
+
+var _ storage = (*memoryStorage)(nil)
+
+func newMemoryStorage() *memoryStorage {
+	return &memoryStorage{
+		content: make(map[string]pdata.ResourceSpans),
+	}
+}
+
+func (st *memoryStorage) createOrAppend(traceID pdata.TraceID, trace pdata.ResourceSpans) error {
+	if trace.IsNil() {
+		return errNilResourceSpans
+	}
+
+	sTraceID := traceID.String()
+
+	st.Lock()
+	existing, ok := st.content[sTraceID]
+	if ok {
+		// we have a trace already, append the spans
+		for i := 0; i < trace.InstrumentationLibrarySpans().Len(); i++ {
+			spans := trace.InstrumentationLibrarySpans().At(i)
+			existing.InstrumentationLibrarySpans().Append(&spans)
+		}
+	} else {
+		newTrace := pdata.NewResourceSpans()
+		newTrace.InitEmpty()
+		trace.CopyTo(newTrace)
```

Comment:
Why do you make copies here but append without copying when `existing` is present?
Unless there is a specific reason, I think the whole logic can be unified as follows:
```go
dest, ok := st.content[sTraceID]
if !ok {
  dest := pdata.NewResourceSpans()
  dest.InitEmpty()
  st.content[sTraceID] = dest
}

// copy spans
trace.InstrumentationLibrarySpans().CopyTo(dest.InstrumentationLibrarySpans())
```

Reply:
That's an oversight. It should be doing a copy of the spans. I just changed the code to do it and added a test.